main.go

```diff
@@ -3,117 +3,157 @@ package main
 	go func() {
-		s := <-sigs
+		wg.Add(1)
+		defer wg.Done()
+		interruptChannel := make(chan os.Signal, 1)
+		signal.Notify(interruptChannel, syscall.SIGTERM, syscall.SIGINT)
+		select {
+		case s := <-interruptChannel:
+			logger.Printf("Received %s, exiting\n", s)
+		case <-shutdownChannel:
+			logger.Println("Received shutdown event, exiting")
+		}
+
 		cancel()
-		logger.Println("Received", s)
-		logger.Println("Exiting")
+		if err := srv.Shutdown(context.Background()); err != nil {
+			// Error from closing listeners, or context timeout:
+			logger.Printf("HTTP server shutdown error: %s\n", err)
+		}
 	}()
 
-	_, err := extensionClient.Register(ctx, extensionName)
-	if err != nil {
-		logger.Fatal(err)
-	}
+	processEvents(ctx, logger, extensionClient)
 
-	initialiseExtension(logger)
+	// Once processEvents returns, signal that it's time to shutdown.
+	shutdownChannel <- struct{}{}
 
-	processEvents(ctx, logger)
+	// Ensure we wait for the HTTP server to gracefully shut down.
+	wg.Wait()
+	logger.Println("Graceful shutdown complete")
 }
 
-func initialiseExtension(logger *log.Logger) {
+func runExtension(logger *log.Logger, wg *sync.WaitGroup) (*http.Server, error) {
 	logger.Println("Initialising")
 
-	vaultAddr := os.Getenv("VAULT_ADDR")
+	vaultAddr := os.Getenv(api.EnvVaultAddress)
 	vaultAuthRole := os.Getenv("VAULT_AUTH_ROLE")
 	vaultAuthProvider := os.Getenv("VAULT_AUTH_PROVIDER")
-	vaultIamServerId := os.Getenv("VAULT_IAM_SERVER_ID")	// Optional
+	vaultIAMServerID := os.Getenv("VAULT_IAM_SERVER_ID") // Optional
 
-	configuredSecrets, err := config.ParseConfiguredSecrets()
-	if err != nil {
-		logger.Fatalf("Failed to parse configured secrets to read: %s", err)
+	if vaultAddr == "" || vaultAuthProvider == "" || vaultAuthRole == "" {
+		return nil, errors.New("missing VAULT_ADDR, VAULT_AUTH_PROVIDER or VAULT_AUTH_ROLE environment variables")
 	}
 
-	if vaultAddr == "" || vaultAuthProvider == "" || vaultAuthRole == "" || len(configuredSecrets) == 0 {
-		logger.Fatal("missing VAULT_ADDR, VAULT_AUTH_PROVIDER, VAULT_AUTH_ROLE, or VAULT_SECRET_ environment variables.")
+	client, config, err := vault.NewClient(logger, vaultAuthRole, vaultAuthProvider, vaultIAMServerID)
+	if err != nil {
+		return nil, fmt.Errorf("error getting client: %w", err)
+	} else if client == nil {
+		return nil, fmt.Errorf("nil client returned: %w", err)
 	}
 
-	client, err := vault.NewClient(logger, vaultAuthRole, vaultAuthProvider, vaultIamServerId)
+	err = writePreconfiguredSecrets(client)
 	if err != nil {
-		logger.Fatalf("error getting client: %s", err)
-	} else if client == nil {
-		logger.Fatalf("nil client returned: %s", err)
+		return nil, err
 	}
 
-	if _, err = os.Stat(config.DefaultSecretDirectory); os.IsNotExist(err) {
-		err = os.MkdirAll(config.DefaultSecretDirectory, 0755)
-		if err != nil {
-			logger.Fatalf("Failed to create directory %s: %s", config.DefaultSecretDirectory, err)
-		}
+	ln, err := net.Listen("tcp", "127.0.0.1:8200")
```

Comment:
Can this port ever conflict with anything (i.e. need to be configurable?)

Reply:
Given the context of Lambdas, I figured it was highly unlikely to ever cause conflicts, but it is an easy one to add configuration for if we get that request. Do you think it would be better to add that configuration upfront?

Reply:
(The implied answer to your question is "yes it's possible" 🙂)

Reply:
We can defer.